src/obstacle_extractor.cpp

```diff
@@ -298,6 +307,31 @@ bool ObstacleExtractor::checkSegmentsCollinearity(const Segment& segment, const
           segment.distanceTo(s2.last_point)  < p_max_merge_spread_);
 }
 
+bool ObstacleExtractor::checkSegmentShrinking(const Segment& segment, const tf::StampedTransform& transform) {
+  Segment segment_transformed = segment;
+  if (p_transform_coordinates_) {
+    segment_transformed.first_point = transformPoint(segment.first_point, transform);
+    segment_transformed.last_point = transformPoint(segment.last_point, transform);
+  }  
+  vector<const Point*> to_check_for_colinearity;
+  for (auto segment_list : prev_segments_) {
```

Comment:
auto creates a copy of elements in prev_segments. since each element is a list it is very expensive operation. I recommend doing auto & segment_list which make s a reference to the list. even better you can make it const auto & segment_list...

same comment applies to the line below that goes through segment lists.